### ISSUE TITLE
Istio ingressgateway routing broken with two different gateways with same port name (SDS)

### ISSUE THREAD

**Issue by p4ali** [istio/istio#12500]
**Describe the bug**
Using https/SDS to secure two gateways(in diff namespaces) which has same port name, the ingressgateway only route to the first

**Expected behavior**
Port name should not be necessary unique, since the gateways are in different namespaces, and ingressgateway should route correctly. 
We have tested http, which works correctly and we hope https has same behavior.

**Steps to reproduce the bug**
* Follow https://preliminary.istio.io/docs/tasks/traffic-management/secure-ingress/sds/ to create certs
* apply following yaml file
```
# httpbin & testcurl with sidecar auto injection
---
apiVersion: v1
kind: Namespace
metadata:
  name: foo
  labels:
    name: foo
    istio-injection: enabled
---
apiVersion: v1
kind: Namespace
metadata:
  name: bar
  labels:
    name: bar
    istio-injection: enabled
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: foo
  namespace: foo
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      credentialName: "foo-credential" # must be the same as secret
    hosts:
    - "foo.example.com"
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: bar
  namespace: bar
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      credentialName: "bar-credential" # must be the same as secret
    hosts:
    - "bar.example.com"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: foo
  namespace: foo
spec:
  hosts:
  - "foo.example.com"
  gateways:
  - foo
  http:
  - match:
    - uri:
        prefix: /status
    route:
    - destination:
        port:
          number: 8000
        host: foo
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: bar
  namespace: bar
spec:
  hosts:
  - "bar.example.com"
  gateways:
  - bar
  http:
  - match:
    - uri:
        prefix: /delay
    route:
    - destination:
        port:
          number: 8000
        host: bar

---
apiVersion: v1
kind: Service
metadata:
  name: foo
  namespace: foo
  labels:
    app: foo
spec:
  ports:
  - name: http
    port: 8000
    targetPort: 8000
  selector:
    app: foo

---
apiVersion: v1
kind: Service
metadata:
  name: bar
  namespace: bar
  labels:
    app: bar
spec:
  ports:
  - name: http
    port: 8000
    targetPort: 8000
  selector:
    app: bar

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: foo
  namespace: foo
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: foo
    spec:
      containers:
      - image: docker.io/citizenstig/httpbin
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 8000
        livenessProbe:
          periodSeconds: 10
          initialDelaySeconds: 20
          httpGet:
            path: /status/200
            port: 8000

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: bar
  namespace: bar
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: bar
    spec:
      containers:
      - image: docker.io/citizenstig/httpbin
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 8000
        livenessProbe:
          periodSeconds: 10
          initialDelaySeconds: 20
          httpGet:
            path: /status/200
            port: 8000
```
* call foo service: `curl -v -HHost:foo.example.com   --resolve foo.example.com:443:127.0.0.1   --cacert foo.example.com/2_intermediate/certs/ca-chain.cert.pem   https://foo.example.com:443/status/418` - success
* call bar service: `curl -v -HHost:bar.example.com   --resolve bar.example.com:443:127.0.0.1   --cacert bar.example.com/2_intermediate/certs/ca-chain.cert.pem   https://bar.example.com:443/delay/1` - failed, reported error:
> * Connected to bar.example.com (127.0.0.1) port 443 (#0)
> * ALPN, offering http/1.1
> * Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
> * successfully set certificate verify locations:
> *   CAfile: bar.example.com/2_intermediate/certs/ca-chain.cert.pem
>   CApath: none
> * TLSv1.2 (OUT), TLS header, Certificate Status (22):
> * TLSv1.2 (OUT), TLS handshake, Client hello (1):
> * OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to bar.example.com:443

**Version**
```
$ istioctl version --remote
client version: version.BuildInfo{Version:"1.1.0-rc.4", GitRevision:"d23daa3d242e51b8fff513f3ac86a708e9bcceba", User:"root", Host:"5ceb25bc-4506-11e9-b4f5-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.0-rc.3-18-gd23daa3"}
...
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T22:29:25Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.4", GitCommit:"f49fa022dbe63faafd0da106ef7e05a29721d3f1", GitTreeState:"clean", BuildDate:"2018-12-14T06:59:37Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Installation**
Follow https://preliminary.istio.io/docs/setup/kubernetes/install/helm/#option-2-install-with-helm-and-tiller-via-helm-install

**Environment**
`Ubuntu 16.04.5 LTS   4.15.0-1036-azure   docker://18.6.0`

**Cluster state**
see attached file

**Comment by duderino** [comment]
@philrud is this similar to something you've looked at in the past?

Actually, this is a special P0 - we need to do it asap, but we can't get it into 1.1 because it shipped already

CC @costinm

**Comment by philrud** [comment]
@duderino another team previously reached out regarding a similar/related issue, but their use case was pretty simple: a single gateway with multiple hosts and the same name for secure ports. My understanding was that this is a known limitation with existing workaround: using distinct names for the ports solves the issue. That's what we ended up recommending them. No further investigation was done at that point, so I don't have any additional info besides what's already covered in the comments here: https://github.com/istio/istio/issues/11509

Now as we actually want to get it fixed asap this is a different story of course -- I'll start digging into what the root cause is and what the least intrusive fix could be.

**Comment by JimmyCYJ** [comment]
I can reproduce this issue with file mount based approach. ingress_config_dump_filemount_ingress.txt is the config dump of ingress gateway envoy, it only has one active_dynamic_listener for foo.example.com
[ingress_config_dump_filemount_ingress.txt](https://github.com/istio/istio/files/2968946/ingress_config_dump_filemount_ingress.txt)
This is my yaml file for reproducing this issue.
[foo-bar-filemount-gateway.yaml.txt](https://github.com/istio/istio/files/2968942/foo-bar-filemount-gateway.yaml.txt)

**Comment by philrud** [comment]
yes, I don't believe this has something to do with SDS, just the lack of scoping while generating listener names or something like that

**Comment by duderino** [comment]
@philrud thanks for digging in.  please take it.  ideally this will go into 1.1.1

**Comment by myidpt** [comment]
Thanks folks for prompt investigation.

**Comment by rshriram** [comment]
The gateway port names have to be unique, if they are sharing the same port. Thats the only way we differentiate different RDS blocks. We went through this motion earlier as well. I wouldn't rock this boat unless absolutely necessary.

**Comment by philrud** [comment]
@rshriram @costinm @duderino 
I have drafted a PR for this: https://github.com/istio/istio/pull/12556 (also fixes https://github.com/istio/istio/issues/11509)
This is currently against 1.1, I'll rebase it over 1.1.1 once the branch is cut.

**Comment by myidpt** [comment]
@duderino Shipping with 1.1.1 makes sense. Have we decided when the 1.1.1 release will be?

**Comment by duderino** [comment]
@philrud please use the release-1.1 branch.

**Comment by duderino** [comment]
@philrud https://github.com/istio/istio/pull/12556 has been merged. Anything left to do here?

**Comment by philrud** [comment]
no, this and #11509 should be addressed by that PR

**Comment by duderino** [comment]
Fixed by #12556 which will be in 1.1.1
